### PR TITLE
[espnet3-recipe] Add LibriTTS recipe (for VITS model)

### PR DESCRIPTION
**Pull request by NewGamezzz**
## What did you change?

Add an ESPnet3 LibriTTS recipe for the VITS model (multi-speaker TTS)

---

## Why did you make this change?

This PR adds a new recipe for LibriTTS + VITS to `egs3/`. It introduces three new stages:

- `compute_xvectors` — extracts SpeechBrain/ECAPA-TDNN speaker embeddings.
- `remove_long_short` — filters utterances by duration (matches the espnet2 `tts.sh` behavior).
- `create_token_list` — builds the phoneme/character/BPE token list from a manifest (mirrors `espnet2/bin/tokenize_text.py`).
---

## Is your PR small enough?

19 files, ~2000 lines added.

---

## Additional Context

This PR is currently a draft. I will mark it as ready for review once VITS training finishes and inference is verified end-to-end.

Plan / checklist:

- [x] Implement provider and runner for parallel x-vector computation
- [x] Add the new stages to `src/system.py`
- [x] Add a stage for metric computation using Versa
- [x] Verify all stages end-to-end:
  - [x] compute_xvectors
  - [x] remove_long_short
  - [x] create_token_list
  - [x] collect_stats
  - [x] train (in progress)
  - [x] infer
  - [x] metrics
- [x] Recipe `README.md`

Current result

| Metric | Value |
|----------|----------:|
| UTMOS | 2.8956 |
| DNSMOS Overall | 3.1845 |
| DNSMOS P.808 | 3.7483 |
| Speaker Similarity | 0.3169 |
| WER (%) | 12.43 |
| CER (%) | 7.10 |

## Summary by CodeRabbit

- **New Features**
  - Added a reusable ESPnet3 text-to-speech recipe template with training, inference, evaluation, packaging, upload, and Gradio demo workflows.
  - Added a complete LibriTTS multi-speaker VITS recipe, including dataset preparation, training, speaker conditioning, inference, and quality metrics.
  - Added reference-audio voice conditioning for LibriTTS demos.
  - Added speaker-embedding extraction support using ESPnet, SpeechBrain, and RawNet3.
  - Added GAN-based TTS training support with configurable optimization and scheduling.

- **Documentation**
  - Added workflow guides and Hugging Face model and demo README templates.

### CODE REVIEW

File: espnet3/systems/tts/models/gan_model.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+import torch
+import torch.distributed as dist
+
+from espnet2.train.abs_gan_espnet_model import AbsGANESPnetModel
+from espnet3.components.modeling.lightning_module import ESPnetLightningModule
+from espnet3.components.modeling.optimization_spec import OptimizationStep
+
+
+class GANTTSLightningModule(ESPnetLightningModule):
+    """Adapt ESPnet2 GAN-TTS model outputs to ESPnet3's named optimizer path."""
```
[the remark this answers is not in the document]
Comment (NewGamezzz) [reply espnet3/systems/tts/models/gan_model.py]: done

**Comment by Masao-Someki** [comment]
@NewGamezzz thanks!
It seems we have CI issue in codecov, which means that your unit test is not enough to cover your new code. Please check the codecov and add unit tests!

**Comment by Masao-Someki** [comment]
It seems we still have CI issue in `test/espnet3/systems/tts/test_xvector.py`?
Could you check it?

File: egs3/libritts/tts/conf/inference.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# Merged over egs3/TEMPLATE/tts/conf/inference.yaml, so only the values this
+# recipe changes are listed here. See the template for the shared defaults
+# (recipe_dir, exp_dir, data_dir, inference_dir, parallel, provider, runner).
+
+##########################################################
+#                   DATASET DEFINITION                   #
+##########################################################
+dataset:
+  _target_: espnet3.components.data.data_organizer.DataOrganizer
+  recipe_dir: ${recipe_dir}
```
Comment (Masao-Someki) [inline egs3/libritts/tts/conf/inference.yaml:13]: I think you can move these args to `TEMPLATE/esp2_gan_tts/`

File: egs3/libritts/tts/conf/inference.yaml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      data_src_args:
+        split: test
+        manifest_path: ${data_dir}/manifests_filtered/test.tsv
+        xvector_dir: ${xvector.save_path}/${xvector.spk_embed_tag}_test
+        load_speech: true
+        inference: true
+  preprocessor:
+    _target_: espnet2.train.preprocessor.CommonPreprocessor
+    fs: 22050
+    train: true
```
Comment (Masao-Someki) [inline egs3/libritts/tts/conf/inference.yaml:32]: BTW you can remove this

File: egs3/libritts/tts/conf/metrics.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        - name: speaker
+          model_tag: espnet/voxcelebs12_ecapa_wavlm_joint
+      wav_key:  wav
+      ref_key:  ref
+      text_key: text
+      use_gpu:  true
+    inputs:
+      wav:  wav
+      ref:  ref
+      text: text
```
Comment (Masao-Someki) [inline egs3/libritts/tts/conf/metrics.yaml:47]: I think we can also move this to `TEMPLATE/esp2_gan_tts`?

File: egs3/libritts/tts/conf/training.yaml [hunk 5]
```diff
@@ -0,0 +1,329 @@
+##########################################################
+#             LIBRITTS VITS TRAINING SETTING             #
+##########################################################
+# Merged over egs3/TEMPLATE/tts/conf/training.yaml, so only the values this
+# recipe changes are listed here. See the template for the shared defaults
+# (num_device/num_nodes, recipe_dir, data_dir, exp_dir, stats_dir,
+# inference_dir, parallel, trainer.accelerator/devices/num_nodes/strategy).
+
+exp_tag: train_vits_libritts
+task: espnet2.tasks.gan_tts.GANTTSTask
```
Comment (Masao-Someki) [inline egs3/libritts/tts/conf/training.yaml:10]: Let's use espnet3's task, the copied one

File: egs3/libritts/tts/conf/training.yaml [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+##########################################################
+create_dataset:
+  recipe_dir: ${recipe_dir}
+
+##########################################################
+#                   DATASET DEFINITION                   #
+##########################################################
+dataset:
+  _target_: espnet3.components.data.data_organizer.DataOrganizer
+  recipe_dir: ${recipe_dir}
```
Comment (Masao-Someki) [inline egs3/libritts/tts/conf/training.yaml:23]: I think we can also move this to TEMPLATE/

File: egs3/libritts/tts/conf/training.yaml [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+    - train
+    - valid
+    - test
+
+##########################################################
+#              UTTERANCE FILTERING SETTING               #
+##########################################################
+remove_long_short:
+  min_wav_duration: 1.0
+  max_wav_duration: 20.0
```
Comment (Masao-Someki) [inline egs3/libritts/tts/conf/training.yaml:69]: I think we can make it a default value?

File: espnet3/systems/tts/models/gan_model.py
Comment (Masao-Someki) [inline espnet3/systems/tts/models/gan_model.py:1]: Please at least add one line of docstrings to all private functions.

File: espnet3/systems/tts/models/gan_model.py [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+                f"GAN-TTS '{name}' loss must be a tensor, but got {type(loss)!r}."
+            )
+
+        stats = output.get("stats")
+        if not isinstance(stats, dict):
+            raise AssertionError(
+                f"GAN-TTS '{name}' stats must be a dict, but got {type(stats)!r}."
+            )
+
+        weight = output.get("weight")
```
Comment (Masao-Someki) [inline espnet3/systems/tts/models/gan_model.py:226]: I am not sure if we need this, do you mean stats/weight could be other format?

File: espnet3/systems/tts/gan_trainer.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        super().__init__(
+            model=model,
+            exp_dir=exp_dir,
+            config=trainer_config,
+            best_model_criterion=best_model_criterion,
+        )
+
+
+def build_gan_trainer(training_config, model) -> GANTTSLightningTrainer:
+    """Build the GAN-specific Lightning trainer for TTS."""
```
Comment (Masao-Someki) [inline espnet3/systems/tts/gan_trainer.py:73]: Please add more documents here~!

File: espnet3/systems/tts/system.py
Comment (Masao-Someki) [inline espnet3/systems/tts/system.py:1]: I think you can create a new `esp2_gan_tts` system class here

File: espnet3/systems/tts/xvector_provider.py [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+            return Speech2Embedding.from_pretrained(
+                model_tag=model_tag,
+                model_file=model_file,
+                batch_size=1,
+                dtype="float32",
+                train_config=None,
+            )
+
+        elif toolkit == "speechbrain":
+            from speechbrain.inference.classifiers import EncoderClassifier
```
Comment (Masao-Someki) [inline espnet3/systems/tts/xvector_provider.py:269]: I think we should add a warning here if people doesn't have speechbrain.

File: espnet3/systems/tts/xvector_provider.py [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+
+            return EncoderClassifier.from_hparams(
+                source=pretrained_model,
+                run_opts={"device": device},
+            )
+
+        elif toolkit == "rawnet":
+            import torch
+            from RawNet3 import RawNet3
+            from RawNetBasicBlock import Bottle2neck
```
Comment (Masao-Someki) [inline espnet3/systems/tts/xvector_provider.py:279]: ditto

File: espnet3/systems/tts/xvector_provider.py [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+        xvec_cfg = self.config.get("xvector", None)
+        if xvec_cfg is None:
+            raise RuntimeError(
+                "xvector configuration not found in training_config. "
+                "Please ensure training_config.xvector is set."
+            )
+
+        toolkit = xvec_cfg.get("toolkit", "speechbrain")
+        pretrained_model = xvec_cfg.get(
+            "pretrained_model", "speechbrain/spkrec-ecapa-voxceleb"
```
Comment (Masao-Someki) [inline espnet3/systems/tts/xvector_provider.py:96]: I think if we don't have the value in config, we can use espnet one (so that people don't need to install additional dependency) or raise an error to fill value.

File: espnet3/systems/tts/xvector_provider.py [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+            ```
+        """
+        xvec_cfg = self.config.get("xvector", None)
+        if xvec_cfg is None:
+            raise RuntimeError(
+                "xvector configuration not found in training_config. "
+                "Please ensure training_config.xvector is set."
+            )
+
+        toolkit = xvec_cfg.get("toolkit", "speechbrain")
```
Comment (Masao-Someki) [inline espnet3/systems/tts/xvector_provider.py:94]: why not espnet?

File: espnet3/systems/tts/xvector_provider.py [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+            if xvec_cfg is None:
+                raise RuntimeError(
+                    "xvector configuration not found in training_config. "
+                    "Please ensure training_config.xvector is set."
+                )
+
+            toolkit = xvec_cfg.get("toolkit", "speechbrain")
+            pretrained_model = xvec_cfg.get(
+                "pretrained_model", "speechbrain/spkrec-ecapa-voxceleb"
+            )
```
Comment (Masao-Someki) [inline espnet3/systems/tts/xvector_provider.py:163]: ditto

**Review by Masao-Someki — COMMENTED** [review]
Hi, let's change the system name to `esp2_gan_tts`! So that we can move it to `systems/esp2_gan_tts/`
